BZ-1800842: Adding steps to regenerate etcd certs

bergerhoffer opened the pull request:
https://bugzilla.redhat.com/show_bug.cgi?id=1800842

Preview: http://file.rdu.redhat.com/~ahoffer/2020/BZ-1800842/backup_and_restore/replacing-failed-master.html (internal)

retroflexer commented on backup_and_restore/replacing-failed-master.adoc:
```diff
@@ -18,10 +18,15 @@ To replace a single master host:
 * xref:../backup_and_restore/replacing-failed-master.adoc#restore-remove-failed-master_replacing-failed-master-host[Remove the member from the etcd cluster].
 
+* If the etcd certificates for the master host have expired, xref:../backup_and_restore/replacing-failed-master.adoc#restore-regenerate-etcd-certs_replacing-failed-master-host[regenerate etcd certificates].
+
```
Lines 21 and 23 should be switched:
If the etcd certificates for the master host are valid, xref:../backup_and_restore/replacing-failed-master.adoc#restore-add-master_replacing-failed-master-host[Add the member back to the etcd cluster].

If the node is new without any certificates or certificates are no longer valid, xref:../backup_and_restore/replacing-failed-master.adoc#restore-regenerate-etcd-certs_replacing-failed-master-host[regenerate etcd certificates and add/recover the member back to the etcd cluster].

retroflexer commented on backup_and_restore/replacing-failed-master.adoc:
```diff
@@ -18,10 +18,15 @@ To replace a single master host:
 // Removing a failed master host from the cluster
 include::modules/restore-remove-failed-master.adoc[leveloffset=+1]
 
+// Regenerating etcd certificates
```
Switch this one with L31 and L32. Change the title as: 
Regenerating etcd certificates and recovering/adding the member to the cluster

bergerhoffer replied:
@retroflexer Would it make sense to use "generate" instead of "regenerate" throughout this, since there's a chance it could be being done for the first time?

retroflexer replied:
I agree.

retroflexer commented on modules/restore-regenerate-etcd-certs.adoc:
```diff
@@ -0,0 +1,173 @@
+[id="restore-regenerate-etcd-certs_{context}"]
+= Regenerating etcd certificates
+
+If the etcd certificates have expired for your master host, you must regenerate them before you can add the member back to the etcd cluster.
```
If the node is a new node or the etcd certificates on the node are no longer valid, you must regenerate them before you can add the member back to the etcd cluster.

retroflexer commented on modules/restore-regenerate-etcd-certs.adoc:
```diff
@@ -0,0 +1,173 @@
++
+// TODO: should this be run on one of the other healthy members?
+
+.. Access the original master, and log in to your cluster as a `cluster-admin` user using the following command.
```
Instead of `original master` say, one of the healthy master nodes.

retroflexer commented on modules/restore-regenerate-etcd-certs.adoc:
```diff
@@ -0,0 +1,173 @@
+.. Access the original master, and log in to your cluster as a `cluster-admin` user using the following command.
++
+----
+[core@ip-10-0-143-125 ~]$ oc login https://localhost:6443
```
Needs sudo before the `oc` command

retroflexer commented on modules/restore-regenerate-etcd-certs.adoc:
```diff
@@ -0,0 +1,173 @@
+.. Create the signer Pod using the file that was generated.
++
+----
+[core@ip-10-0-143-125 ~]$ oc create -f assets/manifests/kube-etcd-cert-signer.yaml
```
Needs `sudo` before `oc` command.

retroflexer commented on modules/restore-add-master.adoc:
```diff
@@ -9,9 +9,10 @@ Follow these steps to add a master host back to the etcd cluster. This procedure
-* The IP address of an existing active etcd member.
+* You have access to the cluster as a user with the `cluster-admin` role.
+* You have SSH access to the master host to add to the etcd cluster.
+* You have the IP address of an existing active etcd member.
```
You have SSH access to the one of the healthy master host.

bergerhoffer replied:
@retroflexer Was this comment supposed to be for the new regenerating procedure's prereq list?

bergerhoffer replied:
Looking at the steps in that procedure, I assume they need both access to a healthy master and to the new one, correct?

retroflexer replied:
Yes.

retroflexer commented on modules/restore-regenerate-etcd-certs.adoc:
```diff
@@ -0,0 +1,173 @@
++
+// TODO: right now, the following steps are a straight copy/paste from restoring lost master hosts procedure (step 5)
+
+. Set up a temporary etcd certificate signer service on your master where you have restored etcd.
```
. Set up a temporary etcd certificate signer service on one of the healthy master nodes.

retroflexer commented on modules/restore-regenerate-etcd-certs.adoc:
```diff
@@ -0,0 +1,173 @@
++
+// TODO: are we able to use the add procedure instead of these steps (which use the etcd-member-recover.sh script)? However, there is a step afterward to delete the remove the signer Pod...
+
+.. Access one of the new master hosts, and log in to your cluster as a `cluster-admin` user using the following command.
```
Access new master host to be added to the cluster ...

retroflexer commented on modules/restore-regenerate-etcd-certs.adoc:
```diff
@@ -0,0 +1,173 @@
+LISTEN   0         128                       *:9943                   *:*
+----
+
+. Add the new master hosts to the etcd cluster.
```
`. Add the new master host to the etcd cluster.`
Singular, not plural. We are adding only single lost master.

retroflexer commented on modules/restore-regenerate-etcd-certs.adoc:
```diff
@@ -0,0 +1,173 @@
+Populating template ./assets/tmp/kube-etcd-cert-signer.yaml.stage1
+Tokenized template now ready: ./assets/manifests/kube-etcd-cert-signer.yaml
+----
+<1> The host name of the original master you just restored, where the signer should be deployed.
```
`The host name of the healthy master, where you are attempting to deploy the signer.`

retroflexer commented on modules/restore-regenerate-etcd-certs.adoc:
```diff
@@ -0,0 +1,173 @@
+ETCD_INITIAL_CLUSTER_STATE="existing"
+Starting etcd..
+----
+<1> Specify both the IP address of the original master where the signer server is running, and the etcd name of the new member.
```
<1> Specify both the IP address of the healthy master where the signer server is running, and the etcd name of the new member.

retroflexer commented on modules/restore-regenerate-etcd-certs.adoc:
```diff
@@ -0,0 +1,173 @@
+
+.. Verify that the new master host has been added to the etcd member list.
+
+... Access the original master and connect to the running etcd container.
```
Access the healthy master ...

bergerhoffer commented on modules/restore-regenerate-etcd-certs.adoc:
```diff
@@ -0,0 +1,172 @@
+.Prerequisites
+
+* You have access to the cluster as a user with the `cluster-admin` role.
+* You have SSH access to the master host to add to the etcd cluster.
```
@retroflexer Can you double check these prereqs I added?

retroflexer replied:
Can we say `new master host` to be clear?

bergerhoffer commented:
@retroflexer I addressed most of your feedback whenever you get another chance to take another look. I'll definitely have more updates to make the flow a little nicer, but at least let's get the procedures down.

retroflexer commented on modules/restore-regenerate-etcd-certs.adoc:
```diff
@@ -0,0 +1,170 @@
++
+----
+[core@ip-10-0-156-255 ~]$ oc login https://localhost:6443
+Authentication required for https://localhost:6443 (openshift)
```
Here also the `oc login` should be done with `sudo`.

retroflexer commented on modules/restore-regenerate-etcd-certs.adoc:
```diff
@@ -0,0 +1,170 @@
+In a terminal that has access to the cluster, run the following command:
++
+----
+$ oc delete pod -n openshift-config etcd-signer
```
Add `sudo` here too.

retroflexer commented on modules/restore-regenerate-etcd-certs.adoc:
```diff
@@ -0,0 +1,168 @@
+In a terminal that has access to the cluster, run the following command:
++
+----
+$ sudo oc delete pod -n openshift-config etcd-signer
```
Here `sudo` is not needed (if fact, it will fail).

bergerhoffer commented:
@geliu2016 Can you take a look at this new procedure for replacing an etcd member that doesn't have etcd certs or the certs are no longer valid?

Preview: http://file.rdu.redhat.com/~ahoffer/2020/BZ-1800842/backup_and_restore/replacing-failed-master.html#restore-regenerate-etcd-certs_replacing-failed-master-host

geliu2016 commented:
LGTM, thanks

bergerhoffer commented:
/cherrypick enterprise-4.2
